spark/src/main/java/com/netflix/iceberg/spark/source/IcebergSource.java

```diff
@@ -89,7 +92,11 @@ public DataSourceReader createReader(DataSourceOptions options) {
           .toUpperCase(Locale.ENGLISH));
     }
 
-    return Optional.of(new Writer(table, lazyConf(), format));
+    String dataLocation = options.get(TableProperties.WRITE_NEW_DATA_LOCATION)
+        .orElse(table.properties().getOrDefault(
+            TableProperties.WRITE_NEW_DATA_LOCATION,
+            new Path(new Path(table.location()), "data").toString()));
+    return Optional.of(new Writer(table, lazyConf(), format, dataLocation));
```

Comment:
Instead of adding parameters to Writer whenever a change like this is made, I'd rather pass the options into Writer and handle these there. The `dataLocation` method could do this work instead of moving it outside the Writer class.

Reply:
I think doing options processing from a `Map<String, String>`, inside a constructor, is a bit of an antipattern. Consider for example writing a unit test for this class in the future. If we pass the `Writer` constructor only a `HashMap`, the unit test would have to construct that `HashMap` in a specific way, i.e. knowing what key-value pairs the constructor is expecting.

Perhaps we can have a builder object that acts as a factory that accepts the `Map` and returns the `Writer`. The `Writer` constructor accepts the builder object and copies the set fields on the builder into its own fields.